### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] False-negative with `??=`

### ISSUE THREAD

**Issue by boris-petrov** [typescript-eslint/typescript-eslint#3553]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```ts
const result: { [key: number]: number } = {};
result[1] ??= 1;
```

```js
{
  "rules": {
    '@typescript-eslint/no-unnecessary-condition': ['error', {
      allowConstantLoopConditions: true,
      allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing: false,
    }],
  }
}
```

**Expected Result**

An error of that rule to be reported.

**Actual Result**

No error.

**Additional Info**

```ts
if (result[1] == null) {
  result[1] = 1;
}
```

This does result in an error so the other code should too.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.27.0` |
| `@typescript-eslint/parser`        | `4.27.0` |
| `TypeScript`                       | `4.3.4` |
| `ESLint`                           | `7.28.0` |
| `node`                             | `16.3.0` |

**Comment by bradzacher** [comment]
The rule just doesn't support these operators yet.

**Comment by bradzacher** [comment]
Just to be clear for posterity's sake - the assignment operators are equivalent to the expression `variable operator (variable = value)` eg:

```ts
const result: { [key: number]: number } = {};
result[1] ??= 1;
result[1] ?? (result[1] = 1);

declare let foo: null;
foo ||= null;
foo || (foo = null);

declare let bar: {};
bar &&= 1;
bar && (bar = 1);
```

But all of those cases are supported - so it's a no brainer to add this
[playground](https://typescript-eslint.io/play/#ts=4.9.3&jsx=true&sourceType=module&code=MYewdgzgLgBATgUwgVwDZQFwwN4wNoDWCAnlmMgLYBGCcAumZTXDAL4wC8OrA3AFCIU6PAEY6MAPwSuI-oLRRR4qTAAU84WM4wRASn58AJgmCoAhohioEsAGYgQjVKn72QMAD4eu5Z64eeHmpu2r6o+nxGJuaW1rBUFljYvHwJLABk6TL8aTCZarkyEUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRgA2pETRoHaJAA04bBCz7skIfHgcA7gGFJyfEJYAZDh2I3ps+U1QZ8cRLpGxqbmFgBKCIgAKhwRTADqsgAWHLD4AMp+lGT4AHII8FZJ-ADWyACSANKcFvFJQvjlAIIAtgAiHMwA5ooAZqbIiHr6AL7DYAC6emMjQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by chharvey** [comment]
I think I found a bug when tsconfig’s `exactOptionalPropertyTypes` is on. With this setting, optional properties are treated as nullish when *accessed* but as non-nullish when being *assigned*.

```ts
// tsconfig:compilerOptions.exactOptionalPropertyTypes = true

const obj: { prop?: [number] } = {};

obj.prop;
//  ^? (property) prop?: [number] | undefined

obj.prop = [42];
//  ^? (property) prop?: [number]
```

I have a function that sets the property once and returns it on the first call, but subsequent calls only return it without setting it:
```ts
function setOnce(value: number): [number] {
    return obj.prop ??= [value]; // ! @typescript-eslint/no-unnecessary-condition: "Unnecessary conditional, value is always truthy."
}

setOnce(1); //=> [1]
setOnce(2); //=> [1]
obj.prop.push(3);
setOnce(1);  //=> [1, 3]
```

The problem here is that typescript-eslint thinks `obj.prop` is non-nullish and reports an error, even though it is nullish before being assigned.

I’m not sure what the right solution is. We want to keep `--exactOptionalPropertyTypes` on to avoid statements like `obj.prop = undefined;`, but we still want to allow the `??=` assignment operator.

**Comment by Josh-Cena** [comment]
@chharvey Thanks for the report. This seems related to https://github.com/typescript-eslint/typescript-eslint/issues/5344 (an optional property is different from a property with `undefined`). Could you send a new issue?
